File: CS/Scaffolding.ValidationErrors/DepartmentContextDataModel/Runtime/DepartmentContextUnitOfWork.cs
using System;
using System.Linq;
using System.Data;
using System.Data.Entity;
using System.Linq.Expressions;
using System.Collections.Generic;
using Scaffolding.ValidationErrors.Common.Utils;
using Scaffolding.ValidationErrors.Common.DataModel;
using Scaffolding.ValidationErrors.Common.DataModel.EntityFramework;
using Scaffolding.ValidationErrors.Model;

namespace Scaffolding.ValidationErrors.DepartmentContextDataModel {
    public class DepartmentContextUnitOfWork : DbUnitOfWork<DepartmentContext>, IDepartmentContextUnitOfWork {
        Lazy<IDepartmentRepository> departmentsRepository;

        public DepartmentContextUnitOfWork(DepartmentContext context)
            : base(context) {
            departmentsRepository = new Lazy<IDepartmentRepository>(() => new DepartmentRepository(this));
        }
        bool IDepartmentContextUnitOfWork.HasChanges() {
            return Context.ChangeTracker.HasChanges();
        }
        IDepartmentRepository IDepartmentContextUnitOfWork.Departments {
            get { return departmentsRepository.Value; }
        }
    }
}

File: VB/Scaffolding.ValidationErrors/DepartmentContextDataModel/Runtime/DepartmentContextUnitOfWork.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Linq
Imports System.Data
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports Scaffolding.ValidationErrors.Common.Utils
Imports Scaffolding.ValidationErrors.Common.DataModel
Imports Scaffolding.ValidationErrors.Common.DataModel.EntityFramework
Imports Scaffolding.ValidationErrors.Model

Namespace Scaffolding.ValidationErrors.DepartmentContextDataModel
	Public Class DepartmentContextUnitOfWork
		Inherits DbUnitOfWork(Of DepartmentContext)
		Implements IDepartmentContextUnitOfWork
		Private departmentsRepository As Lazy(Of IDepartmentRepository)

		Public Sub New(ByVal context As DepartmentContext)
			MyBase.New(context)
			departmentsRepository = New Lazy(Of IDepartmentRepository)(Function() New DepartmentRepository(Me))
		End Sub
		Private Function HasChanges() As Boolean Implements IDepartmentContextUnitOfWork.HasChanges
			Return Context.ChangeTracker.HasChanges()
		End Function
		Private ReadOnly Property Departments() As IDepartmentRepository Implements IDepartmentContextUnitOfWork.Departments
			Get
				Return departmentsRepository.Value
			End Get
		End Property
	End Class
End Namespace
